src/platform/nm-linux-platform.c

```diff
@@ -5935,16 +5950,25 @@ link_set_sriov_params (NMPlatform *platform,
 		return FALSE;
 	}
 
-	if (!nm_platform_sysctl_set (NM_PLATFORM_GET,
-	                             NMP_SYSCTL_PATHID_NETDIR (dirfd,
-	                                                       ifname,
-	                                                       "device/sriov_numvfs"),
-	                             nm_sprintf_buf (buf, "%u", num_vfs))) {
-		_LOGW ("link: couldn't set SR-IOV num_vfs to %d: %s", num_vfs, strerror (errno));
-		return FALSE;
+	if (current_num == 0 && num_vfs == 0) {
+		if (callback) {
+			packed = nm_utils_user_data_pack (g_object_ref (platform),
+			                                  callback,
+			                                  data);
+			nm_utils_invoke_on_idle (sriov_idle_cb, packed, cancellable);
+		}
+		return TRUE;
 	}
 
-	return TRUE;
+	i = 0;
+	if (current_num != 0)
+		values[i++] = "0\n";
+	if (num_vfs != 0)
+		values[i++] = nm_sprintf_bufa (128, "%u\n", num_vfs);
+	values[i++] = NULL;
+
+	nm_sprintf_buf (buf, "/sys/class/net/%s/device/sriov_numvfs", ifname);
```

Comment:
this open a renaming race, that `NMP_SYSCTL_PATHID_NETDIR()` avoided.

Reply:
Yes, but from what I see GLib only allows creating a GFileIOStream
from a GFile, and GFiles can only be constructed by filename, not by file
descriptor. I researched alternative ways to perform asynchronous I/O
on regular files and so far the GLib is the easiest to integrate with.

Alternatives are:

 - use the glibc POSIX asynchronous I/O API (man 7 aio). It internally
   uses a thread pool to perform the operation and runs a callback in
   one of such threads;

 - implements our own solution using threads, which seems not trivial;

 - or stick with the GLib API, and accept the chance of a race during
   interface rename.

What do you think?

Reply:
other alternative: use glib's threadpool. I suppose that exists, because that is what GFileIOStream uses (I presume).